pypeit/calibrations.py

```diff
@@ -501,52 +502,67 @@ def get_flats(self):
-                                                               pixflat_image_files, dark=self.msdark,
-                                                               bias=self.msbias, bpm=self.msbpm)
-        if stacked_flat is not None:
-            # Create pixelflat and illumination flat from illumination flat stack
-            flatField = flatfield.FlatField(stacked_flat, self.spectrograph,
-                                            self.par['flatfield'], self.slits, self.wavetilts)
-            # Run
-            self.flatimages = flatField.run(show=self.show)
-
-            # Save to Masters
-            self.flatimages.to_master_file(masterframe_filename)
-            # Save slits too, in case they were tweaked
-            self.slits.to_master_file()
+        pixelflatImages, illumflatImages = None, None
+        # Check if the image files are the same
+        pix_is_illum = Counter(illum_image_files) == Counter(pixflat_image_files)
+        if len(pixflat_image_files) > 0:
+            pixel_flat = buildimage.buildimage_fromlist(self.spectrograph, self.det,
+                                                        self.par['pixelflatframe'],
+                                                        pixflat_image_files, dark=self.msdark,
+                                                        bias=self.msbias, bpm=self.msbpm)
+            # Initialise the pixel flat
+            pixelFlatField = flatfield.FlatField(pixel_flat, self.spectrograph,
+                                                 self.par['flatfield'], self.slits, self.wavetilts, self.wv_calib)
+            # Generate
+            pixelflatImages = pixelFlatField.run(show=self.show)
+
+        # Only build illum_flat if the input files are different from the pixel flat
+        if (not pix_is_illum) and len(illum_image_files) > 0:
+            illum_flat = buildimage.buildimage_fromlist(self.spectrograph, self.det,
+                                                        self.par['illumflatframe'],
+                                                        illum_image_files, dark=self.msdark,
+                                                        bias=self.msbias, bpm=self.msbpm)
+            # Initialise the pixel flat
+            illumFlatField = flatfield.FlatField(illum_flat, self.spectrograph,
+                                                 self.par['flatfield'], self.slits, self.wavetilts,
+                                                 self.wv_calib, spat_illum_only=True)
+            # Generate
+            illumflatImages = illumFlatField.run(show=self.show)
+
+        # Merge the illum flat with the pixel flat
+        if pixelflatImages is not None:
+            msgs.info("Merging illumflat parameters into FlatImages")
+            self.flatimages = flatfield.merge(pixelflatImages, illumflatImages)
         else:
-            self.flatimages = flatfield.FlatImages(None, None, None, None)
+            # No pixel flat, but there might be an illumflat
```

Comment:
This is not transparent. This will use the pixelflat generated from the illumination flats as the pixel flat. Since you set spat_illum_only=True, I presume that is an image full of ones, but it is unclear. Please add a comment explaining what is going on here and print out some info to be consistent with the above print statement.  Also, does this fail in the case where I want to use a user defined flat field file via self.flatimages above, but then want to use the illumination function generated from twilight flats? This is actually the case with LRIS, and I think that your code does not owrk.